Updated so that the longPress, moveTo, release to drag conversion res…

ahansen1 opened the pull request:
…pects the duration on the longPress if it exists.

moizjv commented:
Can one of the admins verify this patch?

jlipps commented:
thanks for the code and testcase @ahansen1! @imurchie do we want to get this in 1.5 instead?

imurchie commented:
Looks good. Yes, I'll make sure it gets into 1.5, and also check to see if the same things is needed for iOS.

imurchie commented on test/unit/android-controller-specs.js:
```diff
@@ -3,11 +3,16 @@
 
 describe('android-controller', function () {
+  beforeEach(function() {
+    controller.adb = { getApiLevel: function() {} };
```
IS this needed if you are stubbing it below?

ahansen1 replied:
This was a little wonky, this was my first time playing with sinon.  In the spec adb starts out null on the controller.  We needed to define it here for the stub below to work.  Having taken a second look, I would suggest we update this to be: 

`
    controller.adb = {
      getApiLevel: function(cb) {
        cb();
      }
    };
`

And then actually remove the stubbing below.  It isn't really needed since we're not doing any assertions on it anyways and there isn't an original method to restore.  If you agree I'll go ahead and make that change.

ahansen1 replied:
Or to replace the stubbing below to use this direct definition instead of sinon would work as well.  This would limit the scope of the mocking.

imurchie replied:
Either one of these (are they different?) suggestions would work for me.

imurchie commented on lib/devices/android/android-controller.js:
```diff
@@ -906,6 +906,9 @@ androidController.performTouch = function (gestures, cb) {
           // lollipop takes a little longer to get things rolling
           var duration = apiLevel >= 5 ? 2 : 1;
+          if (longPress.options && longPress.options.duration) {
+            duration = Math.max(longPress.options.duration / 1000, 1);
```
I think the `1` here ought to be `duration`.

ahansen1 commented:
I have updated to address the comments.  I agree, using default duration (based on apiLevel) for a minimum instead of just 1 is better.  I also updated the mocking of getApiLevel.  I ended up removing the beforeEach at the top of the spec and doing mocking with sinon localized to the performTouch scenarios.  Continuing to use sinon made it easy to change the apiLevel returned one of the specs.  I used this in the 'longPress duration less than 1 second' spec to make sure the duration 2 (because apiLevel was returned as 6) was passed.

imurchie commented:
Can you squash this into a single commit, then we can merge! Thanks!

ahansen1 committed:
Updated so that the longPress, moveTo, release to drag conversion respects the duration on the longPress if it exists.

ahansen1 commented:
Squashed.  Thanks for the feedback and the quick turn around on getting this into the code base!

imurchie commented:
Thanks for this!

toejough commented:
was this ever applied for ios?  it doesn't look like it was (at least not in appium 1.5.3).

For `action.long_press(x=from_x, y=from_y, duration=5000).move_to(x=0, y=-300).release().perform()` with python client version 0.22... 

the app screen gets flung instead of slowly scrolled and...

my server log shows a time of 0.2s being used as a duration:
```
[HTTP] --> POST /wd/hub/session/b402657b-061a-4ecb-98a4-b3b052f24144/touch/perform {"sessionId":"b402657b-061a-4ecb-98a4-b3b052f24144","actions":[{"options":{"x":134,"duration":5000,"y":437},"action":"longPress"},{"options":{"x":0,"y":-300},"action":"moveTo"},{"options":{},"action":"release"}]}
[MJSONWP] Calling AppiumDriver.performTouch() with args: [[{"options":{"x":134,"dura...
[debug] [iOS] Executing iOS command 'performTouch'
[debug] [UIAuto] Sending command to instruments: target.touch([{"touch":[{"x":134,"y":437}],"time":0.2},{"touch":[{"x":134,"y":137}],"time":0.4}])
[debug] [Instruments] [INST] 2016-07-11 20:57:17 +0000 Debug: evaluation finished
[debug] [Instruments] [INST] 2016-07-11 20:57:17 +0000 Debug: responding with:
[debug] [Instruments] [INST] 2016-07-11 20:57:17 +0000 Debug: Running system command #69: /usr/local/Cellar/node/5.1.0/bin/node /usr/local/lib/node_modules/appium/node_modules/appium-ios-driver/node_modules/appium-uiauto/build/lib/bin/command-proxy-client.js /var/folders/0j/ln7spg1x4dq7blnvvwrq1_2w0000gn/T/instruments_sock 2,{"status":0,"value":""}...
[debug] [Instruments] [INST] 2016-07-11 20:57:38 +0000 Debug: Got new command 69 from instruments: target.touch([{"touch":[{"x":134,"y":437}],"time":0.2},{"touch":[{"x":134,"y":137}],"time":0.4}])
[debug] [Instruments] [INST] 2016-07-11 20:57:38 +0000 Debug: evaluating target.touch([{"touch":[{"x":134,"y":437}],"time":0.2},{"touch":[{"x":134,"y":137}],"time":0.4}])
[debug] [Instruments] [INST] 2016-07-11 20:57:38 +0000 Debug: target.touch(__NSCFArray)
[debug] [Instruments] [INST] 2016-07-11 20:57:38 +0000 Debug: evaluation finished
       2016-07-11 20:57:38 +0000 Debug: responding with:
[debug] [Instruments] [INST] 2016-07-11 20:57:38 +0000 Debug: Running system command #70: /usr/local/Cellar/node/5.1.0/bin/node /usr/local/lib/node_modules/appium/node_modules/appium-ios-driver/node_modules/appium-uiauto/build/lib/bin/command-proxy-client.js /var/folders/0j/ln7spg1x4dq7blnvvwrq1_2w0000gn/T/instruments_sock 2,{"status":0,"value":""}...
[debug] [UIAuto] Socket data received (25 bytes)
[debug] [UIAuto] Got result from instruments: {"status":0,"value":""}
[MJSONWP] Responding to client with driver.performTouch() result: null
[HTTP] <-- POST /wd/hub/session/b402657b-061a-4ecb-98a4-b3b052f24144/touch/perform 200 1648 ms - 76
```
